### PR TITLE
Implemented referents for the extraction

### PR DESCRIPTION
**Pull request by burdoto**
This pull request closes Issue https://github.com/xdrop/fuzzywuzzy/issues/44

### CODE REVIEW

**Comment by burdoto** [comment]
This is odd, I cannot finish the build on my machine either, but for completely unknown reasons; the reported files stayed completely untouched by me.
grafik

**Comment by xdrop** [comment]
Hey, thanks a lot for this!

The error on your IDE is different from the error on travis. To fix the IDE error you need to go to Project Structure and add a module dependency to the parent `fuzzywuzzy` module on the `diffutils` module as sometimes it's not picked up from the pom automatically.

The error on travis though is a failed test case:
Assertion failed: 

```
assert FuzzySearch.extractAll("goolge", choices, 40).size() == 3
                   |                    |            |      |
                   |                    |            8      false
                   |                    [google, bing, facebook, linkedin, twitter, googleplus, bingnews, plexoogl]
                   [(string: google, score: 83, index: 0), (string: bing, score: 23, index: 1), (string: facebook, score: 29, index: 2), (string: linkedin, score: 29, index: 3), (string: twitter, score: 15, index: 4), (string: googleplus, score: 75, index: 5), (string: bingnews, score: 29, index: 6), (string: plexoogl, score: 43, index: 7)]
```

It seems something broke the threshold filtering in `extractAll`

File: src/me/xdrop/fuzzywuzzy/FuzzySearch.java [hunk 1]
```diff
@@ -445,11 +445,11 @@ public static int weightedRatio(String s1, String s2, StringProcessor stringProc
      * @param cutoff  Keep only scores above cutoff
      * @return A list of the results
      */
-    public static List<ExtractedResult> extractAll(String query, Collection<String> choices, int cutoff) {
+    public static List<ExtractedResult<String>> extractAll(String query, Collection<String> choices, int cutoff) {
 
-        Extractor extractor = new Extractor(cutoff);
+        Extractor<String> extractor = new Extractor<>();
```
Comment (xdrop) [inline src/me/xdrop/fuzzywuzzy/FuzzySearch.java]: You missed the `cutoff` here, that's what's causing the test to fail

**Comment by burdoto** [comment]
Ah yeah, missed that one. D:
I was about 100% sure I got everything!

Should be all fixed now.

**Comment by xdrop** [comment]
Great, looking good! 😃 

My only concern is this contains breaking changes for anyone that relied on `Extractor` directly (instead of using the `FuzzySearch` facade) because some the old method signatures (eg. `Extractor#extractWithoutOrder(String, ArrayList<String>, WeightedRatio)`) are not available anymore.

I'll try and see if there's a way around it, otherwise I'll just do a major version bump as it's not backwards compatible.

**Comment by burdoto** [comment]
I could just as well re-add the old methods back inside. i just felt it was a little more handy to not have them, and completely forgot about backwards compatibility

**Comment by burdoto** [comment]
or, analogue to that, rename this extractor implementation to a second, type bound extractor, while the old stays the same

**Comment by xdrop** [comment]
We could move the type argument from the class `Extractor<T>` to `Extractor` and have it on the method since nothing in that class depends on that type argument. Then we could refactor like this and avoid duplicating code:

```
class Extractor {
	
	/**
	 * Returns the list of choices with their associated scores of similarity in a list
	 * of {@link ExtractedResult}
	 *
	 * @param toStringFunction The ToStringFunction to be applied to all choices.
	 * @param query The query string
	 * @param choices The list of choices
	 * @param func The function to apply
	 * @return The list of results
	 */
	public List<ExtractedResult<String>> extractWithoutOrder(String query, Collection<String> choices, Applicable func) {

	    return extractWithoutOrder(query, choices, ToStringFunction.DEFAULT, func);

	}

	/**
	 * Returns the list of choices with their associated scores of similarity in a list
	 * of {@link ExtractedResult}
	 *
	 * @param query The query string
	 * @param choices The list of choices
	 * @param toStringFunction The ToStringFunction to be applied to all choices.
	 * @param func The function to apply
	 * @return The list of results
	 */
	public <T> List<ExtractedResult<T>> extractWithoutOrder(String query, Collection<T> choices,
	                                                        ToStringFunction<T> toStringFunction, Applicable func) {

	    List<ExtractedResult<T>> yields = new ArrayList<>();
	    int index = 0;

	    for (T t : choices) {

	        String s = toStringFunction.apply(t);
	        int score = func.apply(query, s);

	        if (score >= cutoff) {
	            yields.add(new ExtractedResult<>(t, s, score, index));
	        }
	        index++;
	    }

	    return yields;

	}
}
```

What do you think?

**Comment by burdoto** [comment]
That is a good idea as well. i will prepare a commit for that

**Comment by xdrop** [comment]
Awesome. In the meantime you could revert the tests to their original form to make sure they pass as they were.

`git checkout 3481d37221b91a25a680fc5f6484fe0f4e4501e2 -- test`

**Comment by burdoto** [comment]
I think this is how you meant to have it?
Overloaded methods from the Extractor

**Comment by xdrop** [comment]
Perfect. Amazing work, thanks! 👍

**Comment by burdoto** [comment]
Youre very welcome, im happy to help ^^ 
feel free to HMU if ya need anything done, lul

**Comment by burdoto** [comment]
also, maybe this should get a new release, as people (like the guy who asked for this) might find this very useful

**Pull request merged**